source/effect_parser.cpp

```diff
@@ -2666,16 +2666,10 @@ bool reshadefx::parser::parse_variable(type type, std::string name, bool global)
 		{
 			const auto initializer_value = _codegen->emit_load(initializer);
 
-			symbol.id = _codegen->define_variable(location, type, std::move(unique_name), global);
-
 			if (initializer_value != 0)
-			{
 				assert(!global); // Global variables cannot have a dynamic initializer
 
-				expression variable; variable.reset_to_lvalue(location, symbol.id, type);
-
-				_codegen->emit_store(variable, initializer_value);
```

Comment:
This too is necessary for compatibility with SPIR-V. See the comment right above:
```The initializer expression for variables must be a constant```

Reply:
What means "constant"?
Is this code is illegal?
```hlsl
#line 757
	float _1336 = smoothstep(_1334, _1335, discRadius);
```
I means it requies fixes to
```hlsl
#line 757
	float _1336; // avoid set non-constant value
#line 757
	_1336 = smoothstep(_1334, _1335, discRadius);
```

Reply:
This is not about HLSL or GLSL. This is about the SPIRV codegen backend. It requires input in a certain form. One of these requirements is that variables cannot have dynamic initializers (Google for SSA form if you want more information). So yes, this code is illegal in SPIR-V. It's fine in HLSL obviously, but that is not the point. ReShade needs to support all 3 codegen backends.

Reply:
My example is took by current version. (4.1.1.496)
I think that it requires fix for current behavior.

Reply:
I'll say it again. This break SPIR-V. Your changes work fine if you revert modifications to effect_parser.cpp. Please do so, I cannot merge this otherwise.

Reply:
Finally, the problem of initializer with call function, I can not modify it.

Reply:
This is because for https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpFunctionCall the type of the arguments has to equal the type of the function parameters as defined by https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpFunctionParameter. Since the compiler registers those a pointers, the arguments have to be pointers as well. And the only way to obtain a pointer from an expression is to store the value of that expression in a variable first and then get the pointer to the variable. That's why the compiler generates temporary variables for each function argument.
This is not necessary for HLSL/GLSL of course, but the SPIR-V generator shares the same compiler frontend, so it is done for all of them.

Reply:
Thank you for your instruction. It is very useful for learning.